### PR TITLE
Fix Optional

### PR DESCRIPTION
**Pull request by vserret**

### CODE REVIEW

File: src/data_type/mod.rs [hunk 1]
```diff
@@ -1272,7 +1272,6 @@ impl Variant for Union {
         let fields: BTreeSet<String> = self.fields.iter().map(|(f, _)| f.to_string()).collect();
         let other_fields: BTreeSet<String> =
             other.fields.iter().map(|(f, _)| f.to_string()).collect();
-        assert!(fields.is_subset(&other_fields));
```
Comment (vserret) [inline src/data_type/mod.rs:1275]: @ngrislain What is the use of this line ?

File: src/data_type/mod.rs [hunk 2]
```diff
@@ -2296,7 +2312,7 @@ impl Variant for DataType {
                 match (self, other) {
                     // If self and other are from different variants
                     (DataType::Null, _) => true, // Any element of self is also in other
-                    (DataType::Unit(_), DataType::Unit(_)) => true,
+                    (DataType::Unit(_), DataType::Unit(_)) | (DataType::Unit(_), DataType::Optional(_))=> true,
```
Comment (ngrislain) [inline src/data_type/mod.rs:2315]: Can you make sure the formatting is correct (space before =>)
Run `cargo fmt` before you commit.

File: src/data_type/mod.rs (hunk 1 above)
Comment (ngrislain) [reply src/data_type/mod.rs:1275]: I think it is reminiscent of a debugging session.
We should indeed not panic in this case, just return false (cf below)

Comment (ngrislain) [reply src/data_type/mod.rs:1275]: Can you remove this?

File: src/data_type/mod.rs [hunk 3]
```diff
@@ -2709,11 +2725,17 @@ where
...
     type Sum = DataType;
     fn or(self, other: DataType) -> Self::Sum {
-        // Simplify in the case of struct and Unit
-        match self {
-            DataType::Null => other,
-            DataType::Union(u) => u.or(other).into(),
-            s => Union::from_data_type(s).or(other).into(),
+        // Simplify in the case of struct
+        match (self, other) {
+            (DataType::Null, o) => o.clone(),
```
Comment (ngrislain) [inline src/data_type/mod.rs:2730]: Why are you cloning o?
Same below?

File: src/data_type/mod.rs [hunk 4]
```diff
@@ -2709,11 +2725,17 @@ where
 impl Or<DataType> for DataType {
     type Sum = DataType;
     fn or(self, other: DataType) -> Self::Sum {
-        // Simplify in the case of struct and Unit
-        match self {
-            DataType::Null => other,
-            DataType::Union(u) => u.or(other).into(),
-            s => Union::from_data_type(s).or(other).into(),
+        // Simplify in the case of struct
```
Comment (ngrislain) [inline src/data_type/mod.rs:2728]: For consistency only the match on self is done here, the match on the second arg is done in the implementation of the variant.

Comment (ngrislain) [reply src/data_type/mod.rs:2728]: This means `Unit`, `Option` and `Union` should implement `Or` and you should only match `self` against `Unit`, `Option` or `Union` (or non tuple variant such as `Null` or `Any` if relevant) and delegate to the Variant implementation.

**Review by ngrislain — APPROVED** [review]

**Pull request merged**